File: js/sos_service.js
// SOS Emergency Service: Web Audio Alarm Siren, 1-Tap SOS Dispatch & Coordinates Sharing

export class SOSService {
  constructor() {
    this.audioCtx = null;
    this.oscillator = null;
    this.gainNode = null;
    this.sirenInterval = null;
    this.isSirenPlaying = false;
  }

  /**
   * Generates a high-pitch emergency siren using Web Audio API synthesis
   * Works completely offline without needing any audio assets
   */
  toggleSiren() {
    if (this.isSirenPlaying) {
      this.stopSiren();
      return false;
    } else {
      this.startSiren();
      return true;
    }
  }

  startSiren() {
    try {
      const AudioContext = window.AudioContext || window.webkitAudioContext;
      if (!AudioContext) return false;

      this.audioCtx = new AudioContext();
      if (this.audioCtx.state === 'suspended') {
        this.audioCtx.resume();
      }

      this.oscillator = this.audioCtx.createOscillator();
      this.gainNode = this.audioCtx.createGain();

      this.oscillator.type = 'sawtooth';
      this.gainNode.gain.setValueAtTime(0.3, this.audioCtx.currentTime);

      this.oscillator.connect(this.gainNode);
      this.gainNode.connect(this.audioCtx.destination);

      let high = false;
      this.oscillator.frequency.setValueAtTime(700, this.audioCtx.currentTime);

      this.sirenInterval = setInterval(() => {
        if (!this.oscillator || !this.audioCtx) return;
        const now = this.audioCtx.currentTime;
        const targetFreq = high ? 700 : 960;
        this.oscillator.frequency.exponentialRampToValueAtTime(targetFreq, now + 0.25);
        high = !high;
      }, 300);

      this.oscillator.start();
      this.isSirenPlaying = true;
      return true;
    } catch (e) {
      console.error("AudioContext error:", e);
      return false;
    }
  }

  stopSiren() {
    if (this.sirenInterval) {
      clearInterval(this.sirenInterval);
      this.sirenInterval = null;
    }
    if (this.oscillator) {
      try {
        this.oscillator.stop();
        this.oscillator.disconnect();
      } catch (e) {}
      this.oscillator = null;
    }
    if (this.audioCtx) {
      try {
        this.audioCtx.close();
      } catch (e) {}
      this.audioCtx = null;
    }
    this.isSirenPlaying = false;
  }

  /**
   * Generates emergency SOS dispatch text containing real-time location & maps link
   */
  generateEmergencyMessage(location) {
    const mapsLink = `https://maps.google.com/?q=${location.latitude.toFixed(6)},${location.longitude.toFixed(6)}`;
    const address = location.formattedAddress || `${location.district}, ${location.state}`;
    const time = new Date().toLocaleTimeString();

    return `🚨 EMERGENCY SOS! 🚨\nI need immediate emergency help!\nMy current location: ${address}\nLive GPS Map: ${mapsLink}\nTimestamp: ${time}\nPlease send police / ambulance immediately!`;
  }

  /**
   * Launches WhatsApp with pre-filled SOS message
   */
  sendWhatsAppSOS(location) {
    const text = encodeURIComponent(this.generateEmergencyMessage(location));
    window.open(`https://api.whatsapp.com/send?text=${text}`, '_blank');
  }

  /**
   * Launches native SMS app with pre-filled SOS message
   */
  sendSmsSOS(location) {
    const text = encodeURIComponent(this.generateEmergencyMessage(location));
    // Works across iOS and Android
    window.location.href = `sms:?body=${text}`;
  }

  /**
   * Initiates direct phone call
   */
  dial(number) {
    const cleanNumber = number.replace(/[^0-9]/g, '');
    window.location.href = `tel:${cleanNumber}`;
  }
}

File: flutter_app/lib/services/sos_service.dart
import 'package:flutter_phone_direct_caller/flutter_phone_direct_caller.dart';
import 'package:url_launcher/url_launcher.dart';
import 'package:share_plus/share_plus.dart';

class SOSService {
  /// Directly calls the emergency number
  static Future<void> call(String number) async {
    final clean = number.replaceAll(RegExp(r'[^0-9]'), '');
    try {
      bool? res = await FlutterPhoneDirectCaller.callNumber(clean);
      if (res != true) {
        final uri = Uri.parse("tel:$clean");
        if (await canLaunchUrl(uri)) {
          await launchUrl(uri);
        }
      }
    } catch (e) {
      final uri = Uri.parse("tel:$clean");
      if (await canLaunchUrl(uri)) {
        await launchUrl(uri);
      }
    }
  }

  /// Generates the standard SOS distress message
  static String formatEmergencyMessage({
    required String state,
    required String district,
    required double lat,
    required double lng,
  }) {
    final mapsLink = "https://maps.google.com/?q=${lat.toStringAsFixed(6)},${lng.toStringAsFixed(6)}";
    final time = DateTime.now().toLocal().toString().split('.')[0];
    return "🚨 EMERGENCY SOS! 🚨\n"
        "I need immediate emergency help!\n"
        "Location: $district, $state\n"
        "Live GPS Map: $mapsLink\n"
        "Time: $time\n"
        "Please dispatch police / ambulance immediately!";
  }

  /// Sends emergency dispatch via WhatsApp
  static Future<void> sendWhatsAppSOS({
    required String state,
    required String district,
    required double lat,
    required double lng,
  }) async {
    final text = formatEmergencyMessage(state: state, district: district, lat: lat, lng: lng);
    final url = Uri.parse("https://api.whatsapp.com/send?text=${Uri.encodeComponent(text)}");
    if (await canLaunchUrl(url)) {
      await launchUrl(url, mode: LaunchMode.externalApplication);
    } else {
      Share.share(text);
    }
  }

  /// Sends emergency dispatch via SMS
  static Future<void> sendSmsSOS({
    required String state,
    required String district,
    required double lat,
    required double lng,
  }) async {
    final text = formatEmergencyMessage(state: state, district: district, lat: lat, lng: lng);
    final smsUri = Uri.parse("sms:?body=${Uri.encodeComponent(text)}");
    if (await canLaunchUrl(smsUri)) {
      await launchUrl(smsUri);
    } else {
      Share.share(text);
    }
  }

  /// Shares location link via native share sheet
  static Future<void> shareLocation({
    required String state,
    required String district,
    required double lat,
    required double lng,
  }) async {
    final text = formatEmergencyMessage(state: state, district: district, lat: lat, lng: lng);
    await Share.share(text, subject: "🚨 EMERGENCY SOS - Location Coordinates");
  }
}
